### PR TITLE
bazel: Switch over CI

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR enables Bazel builds for almost all of CI. The CI targets that will still be run via Cargo are:

* Cargo Test
* Doc Tests
* Clippy
* Miri Tests
* Cargo Vet Check

All targets that are run via Cargo have a `:rust:` emoji in Buildkite to make them easily identifiable. I have an in-progress PR to run Cargo and Doc tests via Bazel (https://github.com/MaterializeInc/materialize/pull/29266) and there is a path forward for running Clippy and Cargo Vet. The only unknown is running Miri with Bazel but I asked about it in the [Bazel Slack](https://bazelbuild.slack.com/archives/CSV56UT0F/p1725890726657049).

### Motivation

Finishes the M0 of https://github.com/MaterializeInc/materialize/issues/26796

### Tips for reviewer

This PR is split into a few commits which might make it easier to review:

1. Updating all of CI to rely on `bazel-build-<arch>`
2. A few changes to `mzbuild` to better compute the fingerprint
3. Updating the C toolchain to a newly built one that includes sanitizers (full support still in progress)
4. Final changes that button up the build, supporting `tokio_unstable`, pushing Docker images, making sure build info is written.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: ci/nightly/pipeline.template.yml [hunk 1]
```diff
@@ -10,28 +10,39 @@
...
+          - ".bazelrc"
+          - "**/*.bzl"
+        artifact_paths: bazel-explain.log
         depends_on: []
         timeout_in_minutes: 60
-        priority: 50
         agents:
           queue: builder-linux-x86_64
 
-      - id: build-aarch64
```
Comment (def-) [inline ci/nightly/pipeline.template.yml:28]: If we kept the build-aarch64/x86_64 names the diffs would be much smaller.

File: ci/test/pipeline.template.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ dag: true
 
 env:
   CI_BUILDER_SCCACHE: 1
+  CI_BUILD_WITH_BAZEL: 1
```
Comment (def-) [inline ci/test/pipeline.template.yml:19]: Do we also want to set this by default for people building locally?

**Review by def- — APPROVED** [review]
Thanks for triggering nightly, good for me if green

File: ci/test/pipeline.template.yml (hunk 2 above)
Comment (ParkMyCar) [reply ci/test/pipeline.template.yml:19]: Not yet, I want to hold off just a little longer to make sure the C toolchain is stable on developers machines. Probably in a week or so though!

File: ci/nightly/pipeline.template.yml [hunk 3]
```diff
@@ -10,28 +10,39 @@
...
     steps:
       - id: build-x86_64
-        label: Build x86_64
+        label: ":bazel: Build x86_64"
         command: bin/ci-builder run stable bin/pyactivate -m ci.test.build
         inputs:
           - "*"
+          - "WORKSPACE"
+          - ".bazelrc"
+          - "**/*.bzl"
```
Comment (benesch) [inline ci/nightly/pipeline.template.yml:28]: Doesn't `inputs: *` already mean everything? If so, then I don't think you need to explicitly list out the Bazel files. Here and the several jobs below.

**Review by benesch — COMMENTED** [review]
Just about LGTM. Left you one quick comment about the changes to `inputs` in the CI pipelines.

**Review by benesch — APPROVED** [review]

**Pull request merged**